util/rom_chip_info.py

```diff
@@ -11,21 +11,39 @@
 from datetime import datetime
 from pathlib import Path
 
-header_template = r"""
+
+def generate_chip_info_c_source(scm_revision: int, wall_time: str) -> str:
+    """Return the contents of a C source file that defines `kChipInfo`.
+
+    Args:
+        scm_revision: SHA1 sum identifying the current SCM revision.
+        wall_time: A human-readable string containing the build timestamp.
+    """
+
+    SHA1_SUM_SIZE_BYTES = 20
+    SHA1_SUM_SIZE_BITS = SHA1_SUM_SIZE_BYTES * 8
+
+    assert scm_revision.bit_length() <= SHA1_SUM_SIZE_BITS
+
+    # Truncate the SCM revision to the top N bits.
+    SCM_REVISION_SIZE_BITS = 64
+    scm_revision = scm_revision >> (SHA1_SUM_SIZE_BITS - SCM_REVISION_SIZE_BITS)
+
+    return f"""
 // Copyright lowRISC contributors.
 // Licensed under the Apache License, Version 2.0, see LICENSE for details.
 // SPDX-License-Identifier: Apache-2.0
 //
 // --------- W A R N I N G: A U T O - G E N E R A T E D   C O D E !! ---------//
 
-#ifndef _F_CHIPINFO_H__
-#define _F_CHIPINFO_H__
-
-static const char chip_info[128] __attribute__((section(".chip_info"))) =
-  "Version: {%version%}, Build Date: {%build_date%}";
-
-#endif  // _F_CHIPINFO_H__
+#include "sw/device/silicon_creator/lib/chip_info.h"
+#include "hw/ip/lc_ctrl/data/lc_ctrl_regs.h"
 
+const chip_info_t kChipInfo __attribute__((section(".chip_info"))) = {{
+    .scm_revision = {scm_revision:#0x},
+    .hw_rev = (uint64_t)LC_CTRL_HW_REV_REG_RESVAL,
```

Comment:
nit: Can we change the type of `hw_rev` to `lifecycle_hw_rev_t` just so that it's easier to use if we need to access it from ROM/ROM_EXT?
```suggestion
    .hw_rev = (lifecycle_hw_rev_t) {
      .chip_gen = (LC_CTRL_HW_GEN_REG_RESVAL >> LC_CTRL_HW_GEN_CHIP_REV_OFFSET) & LC_CTRL_HW_REV_CHIP_GEN_MASK,
      .chip_rev = (LC_CTRL_HW_REV_REG_RESVAL >> LC_CTRL_HW_REV_CHIP_REV_OFFSET) & LC_CTRL_HW_REV_CHIP_REV_MASK,
    },
```

Reply:
Since this data is already available in a register, is there value in duplicating it in chip_info?

Suppose we build A0, then correct some minor hw bugs and build A1.  Suppose none of the bugs corrected from A0 to A1 require changes to the ROM.  Now the value encoded into the ROM is inconsistent with the value reported by the hardware.

Reply:
I think tagging the ROM (and ROM_EXT) with the intended target (hw_rev) could come in handy in chip postmortems. If we have a malfunctioning chip, we may be able to dump the ROM/ROM_EXT over JTAG and check that it's attempting to run a compatible binary. The commit hash alone can't tell us whether the software was built for the correct target.

Reply:
If we go that route, then we need to document the "allowed inconsistency" between the hw register and the ROM value of this field.

Reply:
Ack on the need to document the potential disagreement between the register and the ROM's chip_info. I'll resolve this thread because I've removed the `hw_rev` from this PR.